fish-rust/src/builtins/shared.rs

```diff
@@ -89,17 +123,29 @@
         args.push(arg.as_utfstr());
     }
     let streams = &mut io_streams_t::new(streams);
-    run_builtin(parser.unpin(), streams, args.as_mut_slice(), builtin);
+
+    match run_builtin(parser.unpin(), streams, args.as_mut_slice(), builtin) {
+        None => false,
+        Some(Ok(())) => {
+            *status_code = 0;
+            true
+        }
+        Some(Err(e)) => {
+            *status_code = e.into();
+            false
```

Comment:
a `false` return will be interpreted as "don't update status" which is wrong. That would mean `builtin false` would never change `$status`.

Instead of mixing `Result<(), CommandError>` and `bool` we should probably use `Option<Status>` everwhere (plus the equivalent `SharedPtr<Status>` or `SharedPtr<i32>` at the language boundary).

Reply:
Yep, forgot to push that fixup yesterday night.

> Instead of mixing `Result<(), CommandError>` and bool we should probably use `Option<Status>` everwhere

The builtins' internals often use short-circuiting on `STATUS_CMD_OK`, as discussed above, if you naively do that on `Option<Status>`, you get nonsensical semantics.